core-java-modules/core-java-concurrency-advanced-3/src/test/java/com/baeldung/exchanger/ExchangerPipeLineUnitTest.java

```diff
@@ -0,0 +1,83 @@
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.Exchanger;
+import java.util.stream.IntStream;
+import org.junit.Test;
+
+public class ExchangerPipeLineUnitTest {
+
+    private static final int BUFFER_SIZE = 100;
+
+    @Test
+    public void givenData_whenPassedThrough_thenCorrect() throws InterruptedException {
+
+        Exchanger<Queue<String>> readerExchanger = new Exchanger<>();
+        Exchanger<Queue<String>> writerExchanger = new Exchanger<>();
+
+        Runnable reader = () -> {
+            Queue<String> readerBuffer = new ConcurrentLinkedQueue<>();
+            while (true) {
+                readerBuffer.add(UUID.randomUUID().toString());
+                if (readerBuffer.size() >= BUFFER_SIZE) {
+                    try {
+                        readerBuffer = readerExchanger.exchange(readerBuffer);
+                    } catch (InterruptedException e) {
+                        Thread.currentThread().interrupt();
+                        throw new RuntimeException(e);
+                    }
+                }
+            }
+        };
+
+        Runnable processor = () -> {
+            Queue<String> processorBuffer = new ConcurrentLinkedQueue<>();
+            Queue<String> writterBuffer = new ConcurrentLinkedQueue<>();
+            try {
+                processorBuffer = readerExchanger.exchange(processorBuffer);
+                while (true) {
+                    writterBuffer.add(processorBuffer.poll());
+                    if (processorBuffer.isEmpty()) {
+                        try {
+                            processorBuffer = readerExchanger.exchange(processorBuffer);
+                            writterBuffer = writerExchanger.exchange(writterBuffer);
+                        } catch (InterruptedException e) {
+                            Thread.currentThread().interrupt();
+                            throw new RuntimeException(e);
+                        }
+                    }
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+        };
+
+        Runnable writer = () -> {
+            Queue<String> writterBuffer = new ConcurrentLinkedQueue<>();
+            try {
+                writterBuffer = writerExchanger.exchange(writterBuffer);
+                while (true) {
+                    System.out.println(writterBuffer.poll());
+                    if (writterBuffer.isEmpty()) {
+                        writterBuffer = writerExchanger.exchange(writterBuffer);
+                    }
+                }
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+        };
+
+        new Thread(reader).start();
+        new Thread(processor).start();
+        new Thread(writer).start();
+        //Thread.currentThread().join();
```

Comment:
This is just an instant deadlock, please remove it

Reply:
Done fixed, replaced with:
CompletableFuture.allOf(runAsync(reader), runAsync(processor), runAsync(writer)).get();

**But left that is commented for the sake build pipeline to stop, else this would run infinitely. To have the test working uncomment the line and execute.**